### PR TITLE
Utilise asyncio more

### PR DESCRIPTION
**Pull request by TRManderson**
Most async scripts were using `bot.post_message` synchronously, which would block the thread running tasks.

This PR introduces `bot.as_async`, which is a fixed version of `AsyncBotWrapper` that runs all functions using the event loop's threadpool executor, and provides an async `api` property.

### CODE REVIEW

**Comment by mitchmcdee** [comment]
Without checking, how does this differ to `bot.run_async`?

**Comment by TRManderson** [comment]
The API just looks way more natural basically, go check out the changes in one of the scripts that used run async

**Comment by mitchmcdee** [comment]
do we need `run_async` anymore then? can it be removed?

**Comment by TRManderson** [comment]
Still need it for requests.get

File: uqcsbot/base.py [hunk 1]
```diff
@@ -260,9 +263,17 @@ def __init__(self, bot: UQCSBot):
 
     def __getattr__(self, name):
         attr = getattr(self.bot, name)
-        if not callable(attr) or not asyncio.iscoroutinefunction(attr):
+        if not callable(attr) or asyncio.iscoroutinefunction(attr):
```
Comment (mitchmcdee) [inline uqcsbot/base.py]: what was the reason for removing the `not` here?

File: uqcsbot/base.py [hunk 2]
```diff
@@ -139,8 +139,8 @@ def api(self):
         return APIWrapper(self.client)
 
     @property
-    def async(self):
-        return AsyncBotWrapper(self.client)
+    def as_async(self):
+        return AsyncBotWrapper(self)
```
Comment (mitchmcdee) [inline uqcsbot/base.py]: why the swap from client to self?

Comment (TRManderson) [reply uqcsbot/base.py]: It's a bot wrapper, not a client wrapper. Having `self.client` there was always a bug.

File: uqcsbot/base.py (hunk 1 above)
Comment (TRManderson) [reply uqcsbot/base.py]: Fixing the behaviour - if it's a coroutine function we want to return it straight up because it's already async

**Comment by TRManderson** [comment]
Nope, user has to use `as_async`

**Comment by mitchmcdee** [comment]
A bit concerned about the complexity and readability sacrifices we're making in exchange for async; ideally that's why I was asking whether it was feasible to enforce async (#282) for everything so that we can abstract things such as with `as_async` as well as simplify the core code base

I'll merge, but I think its something we should think about and discuss more moving forward

**Pull request merged**